### DISCUSSION TITLE
Impact of Incremental mode

### DISCUSSION THREAD

**Discussion by recoules** [bitwuzla/bitwuzla#10] [General]
The documentation says:
```
Enabling this option turns off some optimization techniques.
```
Could you detail more about the advantage and drawbacks of incremental mode? I will certainly do some benchmarks, but if you already have some insight on what could work pretty well or what would be a disaster, I am more than interested.

**Comment by mpreiner** [comment]
When incremental is enabled we disable two preprocessing passes that try to eliminate variables and one to eliminate functions (lambda terms). One of the passes is unconstrained simplification that relies on the fact that a variable only occurs once in a formula, which you can't guarantee when you have multiple check-sat calls.

As for when to use incremental: It depends on the problems you have. If you have problems with multiple check-sat calls, where you can reuse some work from previous check-sat calls then you should use incremental mode. If you would need to reset the solver after each check-sat call you should disable incremental mode.

**Comment by recoules** [comment]
Thank you for these precisions. Would you say that `push`/`pop` based approach is equivalent, better or worse than `assuming` approach?

**Reply by mpreiner** [reply]
For Bitwuzla it shouldn't make any difference since with push/pop we still use assumptions internally.

**Reply by recoules** [reply]
Ok, good to know. Thank you

**Comment by mpreiner** [comment]
This is not relevant anymore since Bitwuzla is by default now incremental and no techniques are disabled anymore.
